h/auth.py

```diff
@@ -0,0 +1,231 @@
+# -*- coding: utf-8 -*-
+"""
+OAuth integration.
+
+Supported grant types
+---------------------
+
+- A client credentials grant using the session and authenticating the client
+  with cross-site request forgery tokens.
+
+Supported token types
+---------------------
+
+- JSON Web Tokens [jwt]_ as bearer tokens.
+
+.. [jwt] https://tools.ietf.org/html/draft-ietf-oauth-json-web-token
+
+Limitations
+-----------
+
+- No support for 3rd party clients exists yet.
+- No support for scopes yet.
+
+"""
+import datetime
+
+import jwt
+from oauthlib.common import generate_client_id
+from oauthlib.oauth2 import RequestValidator as _RequestValidator
+from pyramid.exceptions import BadCSRFToken
+from pyramid import session
+
+from pyramid.authentication import RemoteUserAuthenticationPolicy
+from pyramid.authentication import SessionAuthenticationPolicy
+from pyramid.authorization import ACLAuthorizationPolicy
+from pyramid.util import action_method
+
+from .interfaces import IClientFactory
+
+LEEWAY = 240  # allowance for clock skew in verification
+
+
+class RequestValidator(_RequestValidator):
+
+    """Validates JSON Web Tokens."""
+
+    def authenticate_client(self, request):
+        if request.client_id is None:
+            try:
+                session.check_csrf_token(request, token='assertion')
+            except BadCSRFToken:
+                return False
+            client_id = request.registry.settings['h.client_id']
+            client = get_client(request, client_id)
+            request.client = client
+            request.user = request.session.get('userid')
+        elif request.client_secret is not None:
+            client_id = request.client_id
```

Comment:
I'm not sure I understand when this branch would execute. Where (other than the tests) is `request.client_id` or `request.client_secret` set?